### PR TITLE
APP-792: Refactor dynamic link logic and handle new path "insurance"

### PR DESCRIPTION
**Pull request by hugokallstrom**
### Checklist

- [ ] Functionality is covered by an integration test
- [x] Functionality is accessible in engineering mode

### CODE REVIEW

File: app/src/main/java/com/hedvig/app/SplashActivity.kt [hunk 1]
```diff
@@ -47,88 +52,7 @@ class SplashActivity : BaseActivity(R.layout.activity_splash) {
 
     override fun onNewIntent(intent: Intent) {
         super.onNewIntent(intent)
-        handleFirebaseDynamicLink(intent, null)
```
Comment (StylianosGakis) [inline app/src/main/java/com/hedvig/app/SplashActivity.kt:50]: Before, we were using the intent coming from `onNewIntent` to pass to `FirebaseDynamicLinks.getInstance().getDynamicLink(intent)` . Now we are using `Activity.getIntent()` instead since we don't pass it from here, but the documentation for it says "Return the intent that started this activity." but in the case where we get the deep link from onNewIntent() the one that started the activity wouldn't be the same as the one passed inside the onNewIntent() right?

Would this be a bug in case the activity is already open get get a new intent with a deep link, or does the Intent returned from `Activity.getIntent()` get updated after the `onNewIntent` happens? The docs on `onNewIntent` seem to mention that "Note that getIntent still returns the original Intent." so I just looking at the code makes me think this would make these links not work if the activity is already open 🤔 What do you think? @hugokallstrom

File: app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import com.hedvig.app.feature.settings.Market
+import com.hedvig.app.feature.settings.MarketManager
+
+class DynamicLinkHandler(
+    val context: Context,
+    val marketManager: MarketManager,
+    val startDefaultActivity: (LoginStatus?) -> Unit
+) {
+
+    enum class DynamicLink(val path: String) {
```
Comment (StylianosGakis) [inline app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt:22]: I like having the possibilities as an type what we can do an exhaustive when on, good idea 👍

**Review by StylianosGakis — COMMENTED** [review]
Please do take a good look at least on the onNewIntent() comment, on first sight it appears that there may be a problem on the change

File: app/src/main/java/com/hedvig/app/SplashActivity.kt [hunk 3]
```diff
@@ -30,10 +27,18 @@ class SplashActivity : BaseActivity(R.layout.activity_splash) {
     private val loggedInService: LoginStatusService by inject()
     private val marketManager: MarketManager by inject()
     private val binding by viewBinding(ActivitySplashBinding::bind)
+    private lateinit var dynamicLinkHandler: DynamicLinkHandler
```
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:30]: I wanted to pass this activities context as a constructor param, but maybe it makes more sense to pass it as a function param and inject the class through koin.

File: app/src/main/java/com/hedvig/app/SplashActivity.kt (hunk 1 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:50]: I see! I think this was an issue in the original code as well though. 

`handleFirebaseDynamicLink(intent, null)` this will call the function with loginStatus as null, so the checks in `handleReferralsDeepLink()`, `handleDirectDebitDeepLink()` and `handleForeverDeepLink()` will call `startDefaultActivity()` with null loginStatus as parameter. This will finally call `handleFirebaseDynamicLink(intent, loginStatus)` in `navigateToActivity()`. A bit hard to follow but I think thats whats going on 😅

File: app/src/main/java/com/hedvig/app/SplashActivity.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (StylianosGakis) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:30]: Does it need to be this activity context or just any? Koin can provide the application's context by itself. But if we need the activity it's as you said maybe not super straightforward

File: app/src/main/java/com/hedvig/app/SplashActivity.kt (hunk 1 above)
Comment (StylianosGakis) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:50]: Yeah, we don't want it to pass null all the time either.
How about then having `getLoginStatusAndNavigate` and `navigateToActivity` receive a nullable intent parameter? That way if the deep link comes from the onNewIntent it can use the correct one, and if it doesn't we can do like
```kotlin
val intent = intentComingFromParamters ?: this@Activity.intent
dynamicLinkHandler.handleIntent(intent, loginStatus)
```
so that we actually use the correct intent in this case.

File: app/src/main/java/com/hedvig/app/SplashActivity.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:30]: When starting activities using the application context we can get unexpected back stack behaviour - thats why we need the local context

[the remark this answers is not in the document]
Comment (StylianosGakis) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:30]: Right I didn't know about this, interesting 🤔

File: app/src/main/java/com/hedvig/app/SplashActivity.kt (hunk 1 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:50]: Good idea, fixed 👍

File: app/src/main/java/com/hedvig/app/SplashActivity.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt:30]: I agree with Stelios here - we should probably use Koin to initialize this dependency.

File: app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    fun handleIntent(
+        intent: Intent,
+        loginStatus: LoginStatus?
+    ) {
+        FirebaseDynamicLinks.getInstance()
+            .getDynamicLink(intent)
+            .addOnSuccessListener { pendingDynamicLinkData ->
+                // This will actually be null in some cases
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt:36]: I think this has been tagged with `@RecentlyNonNull`, which has made this comment a lie 😄

File: app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                if (pendingDynamicLinkData != null && pendingDynamicLinkData.link != null) {
+                    val link = pendingDynamicLinkData.link
+                    when (link?.pathSegments?.getOrNull(0)) {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt:39]: Perhaps the `DynamicLink`-enum could own this code, perhaps with a `DynamicLink::from`-function on the `companion object`?

**Review by hedvigoscar — APPROVED** [review]
Approved with nits

File: app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt (hunk 4 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt:36]: Removing the check :)

File: app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt (hunk 5 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt]: Yes, fixed

File: app/src/main/java/com/hedvig/app/SplashActivity.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/SplashActivity.kt]: I went back to the drawing board. Refactored to use extension functions instead and passing the required parameters. Much cleaner IMO.

File: app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt (hunk 4 above)
Comment (hedvigoscar) [reply app/src/main/java/com/hedvig/app/service/DynamicLinkHandler.kt]: Very rude of google to add that annotation in that case 😞

File: app/src/main/java/com/hedvig/app/service/DynamicLink.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
+package com.hedvig.app.service
+
+import android.content.Intent
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/service/DynamicLink.kt:3]: `ktlint -F`

**Review by hedvigoscar — APPROVED** [review]
Approved with nit
